### ISSUE TITLE
Support for "source-based" coverage

### ISSUE THREAD

**Issue by davidhewitt** [lemurheavy/coveralls-public#1502]
Rust (and other LLVM-based languages) can output "source-based coverage" using llvm's profiling instrumentation instead of line-based coverage.

This provides richer feedback to the developer about exactly which parts of a line ran - e.g. for a boolean `&&` can see whether it was always short-circuited.

Rust blog post here: https://blog.rust-lang.org/inside-rust/2020/11/12/source-based-code-coverage.html

As far as I can tell, `coveralls` doesn't have any support for this mode?

**Comment by afinetooth** [comment]
Hi @davidhewitt.

I can't tell from the [article you cite](https://blog.rust-lang.org/inside-rust/2020/11/12/source-based-code-coverage.html) what the full spec of that mode's coverage report format would be, but I can tell you that coveralls supports [LCOV](http://ltp.sourceforge.net/coverage/lcov.php) as a universal coverage report format, and that furthermore, while [LCOV format supports _statement (line)_, _branch_ and _function_ coverage](http://ltp.sourceforge.net/coverage/lcov.php), coveralls currently supports _statement (line)_ and _branch_ coverage. 

By _default_ coveralls will merge branch results into your aggregate coverage % if any branch coverage details are present in your report, _unless_ you change this setting in your repo settings page:

![Screen Shot 2020-11-16 at 1 00 33 PM](https://user-images.githubusercontent.com/918667/99307690-b9175f00-280b-11eb-892f-6b923e54454e.png)

If you can provide more by way of spec on the specific source-based coverage report you're looking to generate, I can get a more specific answer on what aspects we'll support.

Additionally, if we can determine the delta between what we support and what you need, I'll be happy to submit a feature / enhancement request.

**Comment by davidhewitt** [comment]
Hi @afinetooth, many thanks for the quick reply!

Reading through the documentation for llvm's profiling tools, it looks like it's possible to export the profile information into either a JSON format or LCOV format. So it sounds like coveralls should have support for what I'd like to do.

https://llvm.org/docs/CommandGuide/llvm-cov.html#llvm-cov-export

However I do note this piece of the docs:

> When exporting JSON, the regions, functions, expansions, and summaries of the coverage data will be exported. When exporting an lcov trace file, the line-based coverage and summaries will be exported.

So if I'm reading that right, there might be an upstream blocker that the llvm tools don't emit branch and function coverage in the LCOV format. I'll try to verify whether this documentation is correct.

As an alternative, if you're interested in coveralls supporting `llvm-cov`'s JSON format, I can try to uncover the specification for it?

**Comment by afinetooth** [comment]
@davidhewitt yes, we're going to be limited for now to the llvm's LCOV export format. It's a shame that export doesn't include branch coverage because we currently support that, but it sounds like it's worth a test to verify either way.

While ultimately all reports are converted to JSON for POSTs to our API, our format is very specific and comes directly from the LCOV format. While it seems like it could be a heavy lift to support additional formats, an example will provide food for thought and evaluation, so feel free to submit and we'll explore it.
